### ISSUE TITLE
TLS 1.0/1.1 and DTLS 1.0 formally deprecated

### ISSUE THREAD

**Issue by halsteaw** [openssl/openssl#15335]
BCP 195 has been formally updated by RFC 8996 to deprecate TLS 1.0/1.1 and DTLS 1.0

Ref: https://www.rfc-editor.org/info/bcp195

In RFC parlance "TLS 1.0/1.1/DTLS 1.0 MUST NOT be used.  Negotiation of TLS 1.0/1.1/DTLS 1.0 from any version of
   TLS MUST NOT be permitted"

Request the following action be added to the future development timeline (or sooner if deemed appropriate):
Separating TLS 1.0/TLS 1.1/DTLS 1.0 from the default build list, and move them the default deprecated list or weak ssl cipher list
Preserve TLS 1.3/DTLS 1.2 in the default build list.

**Comment by mattcaswell** [comment]
In the forthcoming OpenSSL 3.0, TLS 1.0/1.1 and DTLS 1.0 will not work with the default configuration settings. Applications wishing to use them must explicitly lower the security level. This is the entry from CHANGES:

````
 * The security strength of SHA1 and MD5 based signatures in TLS has been
   reduced. This results in SSL 3, TLS 1.0, TLS 1.1 and DTLS 1.0 no longer
   working at the default security level of 1 and instead requires security
   level 0. The security level can be changed either using the cipher string
   with `@SECLEVEL`, or calling `SSL_CTX_set_security_level()`. This also means
   that where the signature algorithms extension is missing from a ClientHello
   then the handshake will fail in TLS 1.2 at security level 1. This is because,
   although this extension is optional, failing to provide one means that
   OpenSSL will fallback to a default set of signature algorithms. This default
   set requires the availability of SHA1.
````

**Comment by paulidale** [comment]
Should we formally deprecate them as well?

**Comment by vdukhovni** [comment]
The actual behaviour is a bit disconcerting.  The error is rather opaque, could it be easier to understand?
```
$ openssl s_client -starttls smtp -tls1 -quiet -connect localhost:25
Can't use SSL_get_servername
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = <...>
verify return:1
00C0C60008000000:error:0A0C0103:SSL routines:tls_process_key_exchange:internal error:ssl/statem/statem_clnt.c:2254:
```
Instead of `internal error`, could there be something more relevant about the security level???

**Comment by kroeckx** [comment]
Also note that -tls1 -cipher AES256-SHA256 will probably work since it avoids SHA1.

**Comment by vdukhovni** [comment]
> Also note that -tls1 -cipher AES256-SHA256 will probably work since it avoids SHA1.

That doesn't work, the cipher in question is TLS 1.2 cipher:
```
$ openssl ciphers -s -v -tls1 AES256-SHA256
$ 

$ openssl ciphers -s -v -tls1_2 AES256-SHA256
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
$ 
```

**Comment by kroeckx** [comment]
I meant AES256-SHA, where we still allow SHA1.

**Comment by vdukhovni** [comment]
> I meant AES256-SHA, where we still allow SHA1.

I see, I think you mean more generally RSA key transport (not DHE or ECDHE) ciphers, where SHA1 is not used to sign the key exchange:
```
$ openssl s_client -quiet -starttls smtp -tls1_1 -cipher "kRSA" -connect $(uname -n):25
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = dnssec-stats.ant.isi.edu
verify return:1
250 CHUNKING
```
Indeed these (AES256-SHA among them) seem to work without `:@SECLEVEL=0`...

**Comment by kroeckx** [comment]
> I see, I think you mean more generally RSA key transport (not DHE or ECDHE) ciphers, where SHA1 is not used to sign the key exchange:

Yes, which comes down to AES256-SHA and AES128-SHA as far as I know.

The NEWS files properly says so.

**Comment by vdukhovni** [comment]
> Yes, which comes down to AES256-SHA and AES128-SHA as far as I know. The NEWS files properly says so.

I see, yes AES or perhaps CAMELLIA:
```
$ openssl ciphers -tls1_1 -s -v 'kRSA'
AES256-SHA                     SSLv3   Kx=RSA      Au=RSA   Enc=AES(256)               Mac=SHA1
CAMELLIA256-SHA                SSLv3   Kx=RSA      Au=RSA   Enc=Camellia(256)          Mac=SHA1
AES128-SHA                     SSLv3   Kx=RSA      Au=RSA   Enc=AES(128)               Mac=SHA1
CAMELLIA128-SHA                SSLv3   Kx=RSA      Au=RSA   Enc=Camellia(128)          Mac=SHA1

$ openssl ciphers -tls1_1 -s -v 'kRSA:!COMPLEMENTOFALL:@SECLEVEL=0'
AES256-SHA                     SSLv3   Kx=RSA      Au=RSA   Enc=AES(256)               Mac=SHA1
CAMELLIA256-SHA                SSLv3   Kx=RSA      Au=RSA   Enc=Camellia(256)          Mac=SHA1
AES128-SHA                     SSLv3   Kx=RSA      Au=RSA   Enc=AES(128)               Mac=SHA1
CAMELLIA128-SHA                SSLv3   Kx=RSA      Au=RSA   Enc=Camellia(128)          Mac=SHA1
```

**Comment by halsteaw** [comment]
(Not in regards to current conversation)
Addendum to OP for deprecation:

Something to keep an eye on:
https://datatracker.ietf.org/doc/draft-ietf-tls-deprecate-obsolete-kex/

Still draft, but seem to be on its way to be finalized.

**Comment by BugOfBugs** [comment]
>  (AES256-SHA among them) seem to work without `@SECLEVEL=0`

It is an illusion.
All ciphers of TLS1 have `Mac=SHA1`.
All **fail**, if **server** runs with `@SECLEVEL=1:...`, or compiled with `OPENSSL_TLS_SECURITY_LEVEL 1`
RSA (_ciphers with Kx=RSA_) succeeds not because of absence of `SHA1`, but because of **bug**: _client_ and _server_ treat **same** cipher **differently**. **Client** does **use** it. **Server rejects**.

What @vdukhovni reports, perhaps is `1080 SSL_R_TLSV1_ALERT_INTERNAL_ERROR`.
It is a **bug** too. Maybe same: discrepancy in treatment of ciphers.
It manifests at `OPENSSL_TLS_SECURITY_LEVEL 1`. But maybe does not relate to security. Maybe it is a coincidence. Maybe it does not suffice to change the name of error. Maybe there is another mistake.

I am not sure what issues the alert - _server_ to _client_, or vice versa. This is a **bug** too. Logs of _s_server_, _s_client_ are not clear where it comes from.

In log of _s_server_ I see ...reason(...)... as follows:
`@SECLEVEL=1` - 118 `SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM` - in _s_client_ `@SECLEVEL` may be any, 0, or 1.
`@SECLEVEL=0` - 1080 `SSL_R_TLSV1_ALERT_INTERNAL_ERROR` - in _s_client_ `@SECLEVEL=1`

**Comment by 1268** [comment]
note @BugOfBugs  is a known spammer, please ignore

**Comment by BugOfBugs** [comment]
@1268 is a known **vandal**. Destroys knowledge. Burns books. **Spams** many topics with meaningless comments that you see here.
Probably it is he, who destroyed my valid comment.
Destroyer hides under generic name @openssl.
Yet, he fails to destroy definite rubbish of @1268.
Keep @1268 away from making software, and management comments.

I restore my meaningful comment.

>  (AES256-SHA among them) seem to work without `@SECLEVEL=0`

It is an illusion of **client**.
All ciphers of TLS1 have `Mac=SHA1`.
All still **fail**, if **server** runs `@SECLEVEL=1:...`, or compiled with `OPENSSL_TLS_SECURITY_LEVEL 1`
RSA (_ciphers with Kx=RSA_) succeeds not because of absence of `SHA1`, but because of **bug**: _client_ and _server_ treat **same** cipher **differently**. **Client** does **use** it. **Server rejects**.

What @vdukhovni reports, perhaps is `1080 SSL_R_TLSV1_ALERT_INTERNAL_ERROR`.
It is a **bug** too. Maybe same: discrepancy in treatment of ciphers.
It manifests at `OPENSSL_TLS_SECURITY_LEVEL 1`. But maybe does not relate to security. Maybe it is a coincidence. Maybe it does not suffice to change the name of error. Maybe there is another mistake.

I am not sure what issues the alert - _server_ to _client_, or vice versa. This is a **bug** too. Logs of _s_server_, _s_client_ are not clear where it comes from.

In log of _s_server_ I see `...reason(...)...` as follows:
`@SECLEVEL=1` - 118 `SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM` - in _s_client_ `@SECLEVEL` may be any, 0, or 1.
`@SECLEVEL=0` - 1080 `SSL_R_TLSV1_ALERT_INTERNAL_ERROR` - in _s_client_ `@SECLEVEL=1`

**Comment by nhorman** [comment]
There has been lots of conversation in this issue about things we could potentially do, but I don't see a clear actionable work item that is being proposed.  could someone involved with this issue clearly make a proposal here please?

**Comment by vdukhovni** [comment]
My suggestion would be to NOT remove TLS 1.0/1.1 from the build. There are still many use-cases for these for interoperability with old equipment on local networks, cameras, printers, system consoles, ...  Requiring a custom build of OpenSSL to access these is IMHO too high a bar.

Rather, we could *directly* disallow TLS 1.0 and 1.1 at security level 0, rather than rely on most of the ciphers not working, especially since ending up with RSA key transport is far worse than any problem with TLS 1.0/1.1.

Therefore, having working TLS 1.[01] only at security level 0 would be the simplest approach. A cleaner alternative would be a separate control (the cipher security levels are really about the ciphers, and should not ideally mix concerns). However, a new knob would not easily be accessible to users who want a "permissive" connection, while tweaking the cipher list is something many applications allow the user to do.

To get around that issue, a new knob would need an "SSL_CONF" command that can set/clear it via openssl.cnf, and the user can then use a custom openssl.cnf or change the default one.

Sadly, the above is far from an action plan, just somewhat more concrete options. It is far easier for *applications* to decide to leave the past behind, than for a library to remove functionality needed to work with legacy devices and applications.

**Comment by nhorman** [comment]
@vdukhovni thank you, that does seem like a reasonable plan for forward progress here.
@halsteaw @kroeckx @paulidale @mattcaswell any disagreement?

**Comment by BugOfBugs** [comment]
> directly disallow TLS 1.0 and 1.1 at security level _0_
You mean rather level _1_. _0_ would eliminate them.
You are right. Better **ban** rather than **break**.
Yet, better abolish `OPENSSL_TLS_SECURITY_LEVEL`:
https://github.com/openssl/openssl/issues/19867
Overly long reasons:
https://github.com/openssl/openssl/issues/11702
At least make default _0_.
Then nothing would break. Users would regain control, and avoid surprise with default level _1_.
> RSA key transport is far worse than any problem with TLS 1.0/1.1
Why? How? You mean _RSA_ with _TLS 1.0/1.1_ is double trouble? Why double? Both have but same trouble: weak `Mac=SHA1`.
Is it possible to end up with another, **stronger** transport? If so, **advise** user to use that instead. Or **default**, and **advise** as to the choice.
Can _OpenSSL_ **log** advice? - Maybe not directly. Maybe only application can retrieve it, if it chooses. _OpenSSL_ **does not provide advice**. Make it **new feature**.
Although, all other ciphers, not only _RSA_, with  _TLS 1.0/1.1_ list  **weak** `Mac=SHA1`. You needlessly discriminate _RSA_.
Or are **other** transports **better** even with  `Mac=SHA1`? For example: `Kx=ECDH`.

**Comment by BugOfBugs** [comment]
> directly disallow TLS 1.0 and 1.1 at security level _0_

You mean rather _1_. Level _0_ would eliminate them.
You are right. Better **ban** rather than **break**.
Yet, better abolish `OPENSSL_TLS_SECURITY_LEVEL`:
https://github.com/openssl/openssl/issues/19867
Overly long reasons:
https://github.com/openssl/openssl/issues/11702
At least make default _0_.
Then nothing would break. Users would regain control, and avoid surprise with default level _1_.
> RSA key transport is far worse than any problem with TLS 1.0/1.1

Why? How? You mean _RSA_ with _TLS 1.0/1.1_ is double trouble? Why double? Both have but same trouble: weak `Mac=SHA1`.
Is it possible to end up with another, **stronger** transport? If so, **advise** user to use that instead. Or **default**, and **advise** as to the choice.
Can _OpenSSL_ **log** advice? - Maybe not directly. Maybe only application can retrieve it, if it chooses. _OpenSSL_ **does not provide advice**. Make it **new feature**.
Although, all other ciphers, not only _RSA_, with  _TLS 1.0/1.1_ list  **weak** `Mac=SHA1`. You needlessly discriminate _RSA_.
Or are **other** transports **better** even with  `Mac=SHA1`? For example: `Kx=ECDH`.

**Comment by vdukhovni** [comment]
> @vdukhovni thank you, that does seem like a reasonable plan for forward progress here. @halsteaw @kroeckx @paulidale @mattcaswell any disagreement?

I neglected to note that we already have a suitable SSL_CONF control that is even used by default in some "distro" `openssl.cnf` files, and that is the `MinProtocol`, which can be set to TLS 1.2.  So there's no need to remove TLS 1.0 support, we could perhaps set these in our default `openssl.cnf` file, allowing users to opt-out, or let the users make the choice:
```
MinProtocol TLSv1.2
MinProtocol DTLSv1.2
```
See `SSL_CONF_cmd(3)`.

**Comment by mattcaswell** [comment]
> we could perhaps set these in our default openssl.cnf file

Or perhaps just set it to TLSv1.2 by default in the absence of an explicit MinProtocol setting?

**Comment by vdukhovni** [comment]
> > we could perhaps set these in our default openssl.cnf file
> 
> Or perhaps just set it to TLSv1.2 by default in the absence of an explicit MinProtocol setting?

Sure, the main question then is how much of an API break that's considered to be?  Can this go in a minor release, or does it have to wait till 4.0?

**Comment by mattcaswell** [comment]
Yeah good point. It probably would be considered an API break only suitable for 4.0 to do that, but changing the default config file could be done in a minor release. So perhaps we should be considering both.

**Comment by nhorman** [comment]
So what I'm hearing is:
1) Immediately (3.4) modify the default config file to set the min version for SSL to tls 1.2
2) Remove tls 1.2 support entirely (4.0)

do I have that right?

**Comment by mattcaswell** [comment]
> Remove tls 1.2 support entirely (4.0)

Err, no. Set the min protocol by default to TLSv1.2 in 4.0 (which effectively deactivates TLSv1.1 and below by default, unless there is explicit configuration otherwise).

But perhaps we should run this past OTC?

**Comment by mattcaswell** [comment]
**OTC Question: Should we disable TLSv1.1 and below in some way? If so how?**

**Comment by vdukhovni** [comment]
@nhorman:

* 3.4: Tweak the default config file, to require *at least* TLS 1.2, which typically should not affect users updating the library, unless the distro chooses to adopt the change, perhaps by prompting the user to accept it.
* 4.0: Set the compile-time default minimum version to TLS 1.2, which still leaves the TLS 1.0/1.1 code in the library, but inactive without *explicit* configuration to the contrary (e.g. Postfix tends to set the min version explicitly, which is fine for opportunistic TLS, though many users already choose TLS 1.2 and up).

Both allow legacy use-cases to happen without rebuilding the library, just config changes.

**Comment by BugOfBugs** [comment]
>TLSv1.2 by default in absence of explicit MinProtocol

No. It would be 2nd, **hidden** limit. While there should be only 1.
Would **break** old configuration.
It is logical to assume that **all allowed** by default. Easier than having to search for **hidden** limits.
Especially as those limits cause **misleading** reports of **errors**. E.g. `SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM`, and `SSL_R_CA_MD_TOO_WEAK`. User wastes time to solve mysterious weakness. While actually it means is that **restriction** in **configuration** is **too strong**. [`OPENSSL_TLS_SECURITY_LEVEL` is already such **bug**.](https://github.com/openssl/openssl/issues/15335#issuecomment-1680149982) It already **broke** configuration. You suggest to introduce more.

If you do not like choice by user, or default, you better **log advice**. _OpenSSL_ does not **log** ont its own. Application that calls it does. You could make new feature: **advice**. Application would log it for user to learn.

If you insist on **default limit**, make it explicit `MinProtocol`.
This still may **break configuration**, as user may change entire system, rather than update `openssl.cnf`. Then one may not notice change in `openssl.cnf`.
It is better to teach user in log, or in documentation, README.

**Comment by BugOfBugs** [comment]
>TLSv1.2 by default in absence of explicit MinProtocol

No. It would be 2nd, **hidden** limit. While there should be only 1.
Would **break** old configuration.
It is logical to assume that **all allowed** by default. Easier than having to search for **hidden** limits.
Especially as those limits cause **misleading** reports of **errors**. E.g. `SSL_R_NO_SUITABLE_SIGNATURE_ALGORITHM`, and `SSL_R_CA_MD_TOO_WEAK`. User wastes time to solve mysterious weakness. While actually it means that **restriction** in **configuration** is **too strong**. [`OPENSSL_TLS_SECURITY_LEVEL` is already such **bug**.](https://github.com/openssl/openssl/issues/15335#issuecomment-1680149982) It already **broke** configuration. You suggest to introduce more.

If you do not like choice by user, or default, you better **log advice**. _OpenSSL_ does not **log** ont its own. Application that calls it does. You could make new feature: **advice**. Application would log it for user to learn.

If you insist on **default limit**, make it explicit `MinProtocol`.
This still may **break configuration**, as user may change entire system, rather than update `openssl.cnf`. Then one may not notice change in `openssl.cnf`.
It is better to teach user in log, or in documentation, README.

**Comment by t8m** [comment]
See also https://github.com/openssl/project/issues/672#issuecomment-2194000735

**Comment by BugOfBugs** [comment]
@t8m created **trivial** issue. Should not create issues for everything that works. You may **delete** it, and move part of comment here.

You are right. Another way to impose default is to **compile without** alternatives. E.g.: `no-tls1`.
Would do so eventually, when nobody uses that.
Then indeed you could say "_**not supported**_": _`SSL_R_UNSUPPORTED_PROTOCOL_` in your issue, `SSL_R_UNSUPPORTED_SSL_VERSION` [`0A0C0103` of old OpenSSL](https://github.com/openssl/openssl/issues/15335#issuecomment-1263763868) or `0A000103` of version 3.
Effect would be similar to `MinProtocol`. [You should still tell user, explicitly.](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753)

If you **disable**, you should say: "**disabled**". Lest you confuse user. So, one knows that may enable.
Error that reports to other end (_client, server_) may say "_**not supported**_" anyway, as other end does not control configuration.
"**Disabled**" may still confuse user. One may believe _OpenSSL_ disables something during run. Better specific: "**disabled in configuration**".

As noted in this issue, confusing wording:
"_**not supported**_",
["_**internal error**_"](https://github.com/openssl/openssl/issues/15335#issuecomment-1263763868),
["_**no suitable**_", "_**weak**_"](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753).

**Comment by BugOfBugs** [comment]
@t8m created **trivial** issue. Should not create issues for everything that works. You may **delete** it, and move part of comment here.

You are right. Another way to impose default is to **compile without** alternatives. E.g.: `no-tls1`.
Would do so eventually, when nobody uses that.
Then indeed you could say "_**not supported**_": _`SSL_R_UNSUPPORTED_PROTOCOL_` in your issue, `SSL_R_UNSUPPORTED_SSL_VERSION` [`0A0C0103` of old _OpenSSL_](https://github.com/openssl/openssl/issues/15335#issuecomment-1263763868) or `0A000103` of version 3.
Effect would be similar to `MinProtocol`. [Should still tell user, explicitly.](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753)

If you **disable**, you should say: "_**disabled**_". Lest confuse user. So, one knows that may enable.
Error that reports to other end (_client, server_) may say "_**not supported**_" anyway, as other end does not control configuration.
"_**Disabled**_" may still confuse user. One may believe _OpenSSL_ disables something during run. Better specific: "_**disabled in configuration**_".

As noted in this issue, confusing wording:
"_**not supported**_",
["_**internal error**_"](https://github.com/openssl/openssl/issues/15335#issuecomment-1263763868),
["_**no suitable**_", "_**weak**_"](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753).

**Comment by kroeckx** [comment]
Note that there are people who use the library without a config file. If you want to set a default, it should not be in the config file.

**Comment by BugOfBugs** [comment]
You can not program **specifically** for **every** possible use. It is _oxymoron_.
And you do not. You threaten to impose **implicit** default on all, with excuse that few refuse to use _openssl.cnf_

In any case, default **must** be **explicit**. - _**Advice**_ In configuration, reports of errors, log, return of functions, documentation, configuration.

If user **discards configuration** _openssl.cnf_, this counts as **opt out** of default. Then user has to impose one's own limits.
User still receives _openssl.cnf_. Can read _advice_ in it. Or where I mentioned.

Your mistake is similar to [one in another issue that mistakes demand for specification](https://github.com/openssl/openssl/issues/22203#issuecomment-2174189819).
It is **user**, who decides what to deprecate. [_OpenSSL_ is a tool to enable all.](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753) Should only **advise**.

**Comment by t8m** [comment]
**OTC: We would like to see a concrete proposal how the deprecation would look like.**

Created https://github.com/openssl/project/issues/750 to track the proposal.
